## Exercises - Analytics Functions

Let us take care of the exercises related to analytics functions. We will be using HR database for the same.

* Get all the employees who is making more than average salary with in each department.
* Get cumulative salary for one of the department along with department name.
* Get top 3 paid employees with in each department by salary (use dense_rank)
* Get top 3 products sold in the month of 2014 January by revenue.
* Get top 3 products in each category sold in the month of 2014 January by revenue.

### Prepare HR Database

Here are the steps to prepare HR database.
* Connect to HR DB using `psql` or SQL Workbench. Here is the sample `psql` command.

```shell
psql -h localhost \
    -p 5432 \
    -d itversity_hr_db \
    -U itversity_hr_user \
    -W
```

* Run scripts to create tables and load the data. You can also drop the tables if they already exists.

```sql
\i /data/hr_db/drop_tables_pg.sql
\i /data/hr_db/create_tables_pg.sql
\i /data/hr_db/load_data_pg.sql
```

* Validate to ensure that data is available in the tables by running these queries.

In [1]:
%load_ext sql

In [2]:
%env DATABASE_URL=postgresql://itv001477_hr_user:xndxklzpkk2h37en8z32vzeekbuvwkca@m01.itversity.com:5433/itv001477_hr_db

env: DATABASE_URL=postgresql://itv001477_hr_user:xndxklzpkk2h37en8z32vzeekbuvwkca@m01.itversity.com:5433/itv001477_hr_db


In [4]:
%sql SELECT * FROM employees LIMIT 10

10 rows affected.


[(100, 'Steven', 'King', 'SKING', '515.123.4567', datetime.date(1987, 6, 17), 'AD_PRES', Decimal('24000.00'), None, None, 90),
 (101, 'Neena', 'Kochhar', 'NKOCHHAR', '515.123.4568', datetime.date(1989, 9, 21), 'AD_VP', Decimal('17000.00'), None, 100, 90),
 (102, 'Lex', 'De Haan', 'LDEHAAN', '515.123.4569', datetime.date(1993, 1, 13), 'AD_VP', Decimal('17000.00'), None, 100, 90),
 (103, 'Alexander', 'Hunold', 'AHUNOLD', '590.423.4567', datetime.date(1990, 1, 3), 'IT_PROG', Decimal('9000.00'), None, 102, 60),
 (104, 'Bruce', 'Ernst', 'BERNST', '590.423.4568', datetime.date(1991, 5, 21), 'IT_PROG', Decimal('6000.00'), None, 103, 60),
 (105, 'David', 'Austin', 'DAUSTIN', '590.423.4569', datetime.date(1997, 6, 25), 'IT_PROG', Decimal('4800.00'), None, 103, 60),
 (106, 'Valli', 'Pataballa', 'VPATABAL', '590.423.4560', datetime.date(1998, 2, 5), 'IT_PROG', Decimal('4800.00'), None, 103, 60),
 (107, 'Diana', 'Lorentz', 'DLORENTZ', '590.423.5567', datetime.date(1999, 2, 7), 'IT_PROG', Decimal('4200.00'), None, 103, 60),
 (108, 'Nancy', 'Greenberg', 'NGREENBE', '515.124.4569', datetime.date(1994, 8, 17), 'FI_MGR', Decimal('12000.00'), None, 101, 100),
 (109, 'Daniel', 'Faviet', 'DFAVIET', '515.124.4169', datetime.date(1994, 8, 16), 'FI_ACCOUNT', Decimal('9000.00'), None, 108, 100)]

In [7]:
%%sql 

SELECT * FROM departments 
ORDER BY manager_id NULLS LAST
LIMIT 10

 * postgresql://itv001477_hr_user:***@m01.itversity.com:5433/itv001477_hr_db
10 rows affected.


[(90, 'Executive', 100, 1700),
 (60, 'IT', 103, 1400),
 (100, 'Finance', 108, 1700),
 (30, 'Purchasing', 114, 1700),
 (50, 'Shipping', 121, 1500),
 (80, 'Sales', 145, 2500),
 (10, 'Administration', 200, 1700),
 (20, 'Marketing', 201, 1800),
 (40, 'Human Resources', 203, 2400),
 (70, 'Public Relations', 204, 2700)]

### Prepare Retail Database

Make sure to drop and recreate the tables before taking care of the exercises related to retail database.

* Ensure that we have required database and user for retail data. **We might provide the database as part of our labs.** Here are the instructions to use `psql` for setting up the required database (if required) and tables.

```shell
psql -U postgres -h localhost -p 5432 -W
```

```sql
CREATE DATABASE itversity_retail_db;
CREATE USER itversity_retail_user WITH ENCRYPTED PASSWORD 'retail_password';
GRANT ALL ON DATABASE itversity_retail_db TO itversity_retail_user;
```

* Create Tables using the script provided. You can either use `psql` or **SQL Workbench**.

```shell
psql -U itversity_retail_user \
  -h localhost \
  -p 5432 \
  -d itversity_retail_db \
  -W
```

* You can drop the existing tables.

```sql
DROP TABLE IF EXISTS order_items CASCADE;
DROP TABLE IF EXISTS orders CASCADE;
DROP TABLE IF EXISTS customers CASCADE;
DROP TABLE IF EXISTS products CASCADE;
DROP TABLE IF EXISTS categories CASCADE;
DROP TABLE IF EXISTS departments CASCADE;
```

* Once the tables are dropped you can run below script to create the tables for the purpose of exercises.

```sql
\i /data/retail_db/create_db_tables_pg.sql
```

* Data shall be loaded using the script provided.

```sql
\i /data/retail_db/load_db_tables_pg.sql
```

* Run queries to validate we have data in all the 3 tables.

In [1]:
%load_ext sql

In [2]:
%env DATABASE_URL=postgresql://itversity_retail_user:retail_password@localhost:5432/itversity_retail_db

env: DATABASE_URL=postgresql://itversity_retail_user:retail_password@localhost:5432/itversity_retail_db


In [6]:
%sql SELECT * FROM orders LIMIT 10

10 rows affected.


[(1, datetime.datetime(2013, 7, 25, 0, 0), 11599, 'CLOSED'),
 (2, datetime.datetime(2013, 7, 25, 0, 0), 256, 'PENDING_PAYMENT'),
 (3, datetime.datetime(2013, 7, 25, 0, 0), 12111, 'COMPLETE'),
 (4, datetime.datetime(2013, 7, 25, 0, 0), 8827, 'CLOSED'),
 (5, datetime.datetime(2013, 7, 25, 0, 0), 11318, 'COMPLETE'),
 (6, datetime.datetime(2013, 7, 25, 0, 0), 7130, 'COMPLETE'),
 (7, datetime.datetime(2013, 7, 25, 0, 0), 4530, 'COMPLETE'),
 (8, datetime.datetime(2013, 7, 25, 0, 0), 2911, 'PROCESSING'),
 (9, datetime.datetime(2013, 7, 25, 0, 0), 5657, 'PENDING_PAYMENT'),
 (10, datetime.datetime(2013, 7, 25, 0, 0), 5648, 'PENDING_PAYMENT')]

In [9]:
%sql SELECT count(1) FROM orders

 * postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db
1 rows affected.


[(68883,)]

In [7]:
%sql SELECT * FROM order_items LIMIT 10

 * postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db
10 rows affected.


[(1, 1, 957, 1, 299.98, 299.98),
 (2, 2, 1073, 1, 199.99, 199.99),
 (3, 2, 502, 5, 250.0, 50.0),
 (4, 2, 403, 1, 129.99, 129.99),
 (5, 4, 897, 2, 49.98, 24.99),
 (6, 4, 365, 5, 299.95, 59.99),
 (7, 4, 502, 3, 150.0, 50.0),
 (8, 4, 1014, 4, 199.92, 49.98),
 (9, 5, 957, 1, 299.98, 299.98),
 (10, 5, 365, 5, 299.95, 59.99)]

In [10]:
%sql SELECT count(1) FROM order_items

 * postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db
1 rows affected.


[(172198,)]

In [8]:
%sql SELECT * FROM products LIMIT 10

 * postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db
10 rows affected.


[(1, 2, 'Quest Q64 10 FT. x 10 FT. Slant Leg Instant U', '', 59.98, 'http://images.acmesports.sports/Quest+Q64+10+FT.+x+10+FT.+Slant+Leg+Instant+Up+Canopy'),
 (2, 2, "Under Armour Men's Highlight MC Football Clea", '', 129.99, 'http://images.acmesports.sports/Under+Armour+Men%27s+Highlight+MC+Football+Cleat'),
 (3, 2, "Under Armour Men's Renegade D Mid Football Cl", '', 89.99, 'http://images.acmesports.sports/Under+Armour+Men%27s+Renegade+D+Mid+Football+Cleat'),
 (4, 2, "Under Armour Men's Renegade D Mid Football Cl", '', 89.99, 'http://images.acmesports.sports/Under+Armour+Men%27s+Renegade+D+Mid+Football+Cleat'),
 (5, 2, 'Riddell Youth Revolution Speed Custom Footbal', '', 199.99, 'http://images.acmesports.sports/Riddell+Youth+Revolution+Speed+Custom+Football+Helmet'),
 (6, 2, "Jordan Men's VI Retro TD Football Cleat", '', 134.99, 'http://images.acmesports.sports/Jordan+Men%27s+VI+Retro+TD+Football+Cleat'),
 (7, 2, 'Schutt Youth Recruit Hybrid Custom Football H', '', 99.99, 'http://images.acmesports.sports/Schutt+Youth+Recruit+Hybrid+Custom+Football+Helmet+2014'),
 (8, 2, "Nike Men's Vapor Carbon Elite TD Football Cle", '', 129.99, 'http://images.acmesports.sports/Nike+Men%27s+Vapor+Carbon+Elite+TD+Football+Cleat'),
 (9, 2, 'Nike Adult Vapor Jet 3.0 Receiver Gloves', '', 50.0, 'http://images.acmesports.sports/Nike+Adult+Vapor+Jet+3.0+Receiver+Gloves'),
 (10, 2, "Under Armour Men's Highlight MC Football Clea", '', 129.99, 'http://images.acmesports.sports/Under+Armour+Men%27s+Highlight+MC+Football+Cleat')]

In [11]:
%sql SELECT count(1) FROM products

 * postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db
1 rows affected.


[(1345,)]

### Exercise 1

Get all the employees who is making more than average salary with in each department.

* Use HR database employees and department tables for this problem.
* Compute average salary expense for each department and get those employee details who are making more salary than average salary.
* Make sure average salary expense per department is rounded off to 2 decimals.
* Output should contain employee_id, department_name, salary and avg_salary_expense (derived field).
* Data should be sorted in ascending order by department_id and descending order by salary.

|employee_id|department_name|salary|avg_salary_expense|
|---|---|---|---|
|201|Marketing|13000.00|9500.00|
|114|Purchasing|11000.00|4150.00|
|121|Shipping|8200.00|3475.56|
|120|Shipping|8000.00|3475.56|
|122|Shipping|7900.00|3475.56|
|123|Shipping|6500.00|3475.56|
|124|Shipping|5800.00|3475.56|
|184|Shipping|4200.00|3475.56|
|185|Shipping|4100.00|3475.56|
|192|Shipping|4000.00|3475.56|
|193|Shipping|3900.00|3475.56|
|188|Shipping|3800.00|3475.56|
|137|Shipping|3600.00|3475.56|
|189|Shipping|3600.00|3475.56|
|141|Shipping|3500.00|3475.56|
|103|IT|9000.00|5760.00|
|104|IT|6000.00|5760.00|
|145|Sales|14000.00|8955.88|
|146|Sales|13500.00|8955.88|
|147|Sales|12000.00|8955.88|
|168|Sales|11500.00|8955.88|
|148|Sales|11000.00|8955.88|
|174|Sales|11000.00|8955.88|
|149|Sales|10500.00|8955.88|
|162|Sales|10500.00|8955.88|
|156|Sales|10000.00|8955.88|
|150|Sales|10000.00|8955.88|
|169|Sales|10000.00|8955.88|
|170|Sales|9600.00|8955.88|
|163|Sales|9500.00|8955.88|
|151|Sales|9500.00|8955.88|
|157|Sales|9500.00|8955.88|
|158|Sales|9000.00|8955.88|
|152|Sales|9000.00|8955.88|
|100|Executive|24000.00|19333.33|
|108|Finance|12000.00|8600.00|
|109|Finance|9000.00|8600.00|
|205|Accounting|12000.00|10150.00|

In [25]:
%%sql
select * from (
    select e.employee_id,d.department_name,e.salary
    ,round(AVG(e.salary) over (partition by e.department_id),2) as avg_salary_expense from employees e
    join departments d on e.department_id=d.department_id
    order by e.department_id,e.salary desc
) q
where q.salary > q.avg_salary_expense



 * postgresql://itv001477_hr_user:***@m01.itversity.com:5433/itv001477_hr_db
38 rows affected.


[(201, 'Marketing', Decimal('13000.00'), Decimal('9500.00')),
 (114, 'Purchasing', Decimal('11000.00'), Decimal('4150.00')),
 (121, 'Shipping', Decimal('8200.00'), Decimal('3475.56')),
 (120, 'Shipping', Decimal('8000.00'), Decimal('3475.56')),
 (122, 'Shipping', Decimal('7900.00'), Decimal('3475.56')),
 (123, 'Shipping', Decimal('6500.00'), Decimal('3475.56')),
 (124, 'Shipping', Decimal('5800.00'), Decimal('3475.56')),
 (184, 'Shipping', Decimal('4200.00'), Decimal('3475.56')),
 (185, 'Shipping', Decimal('4100.00'), Decimal('3475.56')),
 (192, 'Shipping', Decimal('4000.00'), Decimal('3475.56')),
 (193, 'Shipping', Decimal('3900.00'), Decimal('3475.56')),
 (188, 'Shipping', Decimal('3800.00'), Decimal('3475.56')),
 (189, 'Shipping', Decimal('3600.00'), Decimal('3475.56')),
 (137, 'Shipping', Decimal('3600.00'), Decimal('3475.56')),
 (141, 'Shipping', Decimal('3500.00'), Decimal('3475.56')),
 (103, 'IT', Decimal('9000.00'), Decimal('5760.00')),
 (104, 'IT', Decimal('6000.00'), Decimal('5760.00')),
 (145, 'Sales', Decimal('14000.00'), Decimal('8955.88')),
 (146, 'Sales', Decimal('13500.00'), Decimal('8955.88')),
 (147, 'Sales', Decimal('12000.00'), Decimal('8955.88')),
 (168, 'Sales', Decimal('11500.00'), Decimal('8955.88')),
 (174, 'Sales', Decimal('11000.00'), Decimal('8955.88')),
 (148, 'Sales', Decimal('11000.00'), Decimal('8955.88')),
 (162, 'Sales', Decimal('10500.00'), Decimal('8955.88')),
 (149, 'Sales', Decimal('10500.00'), Decimal('8955.88')),
 (156, 'Sales', Decimal('10000.00'), Decimal('8955.88')),
 (169, 'Sales', Decimal('10000.00'), Decimal('8955.88')),
 (150, 'Sales', Decimal('10000.00'), Decimal('8955.88')),
 (170, 'Sales', Decimal('9600.00'), Decimal('8955.88')),
 (151, 'Sales', Decimal('9500.00'), Decimal('8955.88')),
 (163, 'Sales', Decimal('9500.00'), Decimal('8955.88')),
 (157, 'Sales', Decimal('9500.00'), Decimal('8955.88')),
 (158, 'Sales', Decimal('9000.00'), Decimal('8955.88')),
 (152, 'Sales', Decimal('9000.00'), Decimal('8955.88')),
 (100, 'Executive', Decimal('24000.00'), Decimal('19333.33')),
 (108, 'Finance', Decimal('12000.00'), Decimal('8600.00')),
 (109, 'Finance', Decimal('9000.00'), Decimal('8600.00')),
 (205, 'Accounting', Decimal('12000.00'), Decimal('10150.00'))]

### Exercise 2

Get cumulative salary with in each department for Finance and IT department along with department name.

* Use HR database employees and department tables for this problem.
* Compute cumulative salary expense for **Finance** as well as **IT** departments with in respective departments.
* Make sure cumulative salary expense per department is rounded off to 2 decimals.
* Output should contain employee_id, department_name, salary and cum_salary_expense (derived field).
* Data should be sorted in ascending order by department_name and then salary.

|employee_id|department_name|salary|cum_salary_expense|
|---|---|---|---|
|113|Finance|6900.00|6900.00|
|111|Finance|7700.00|14600.00|
|112|Finance|7800.00|22400.00|
|110|Finance|8200.00|30600.00|
|109|Finance|9000.00|39600.00|
|108|Finance|12000.00|51600.00|
|107|IT|4200.00|4200.00|
|106|IT|4800.00|9000.00|
|105|IT|4800.00|13800.00|
|104|IT|6000.00|19800.00|
|103|IT|9000.00|28800.00|

In [3]:
%%sql
select count(1) from employees

1 rows affected.


[(109,)]

In [4]:
%%sql
select e.employee_id,d.department_name,e.salary,round(sum(e.salary) over(partition by d.department_name
    order by e.salary
    ROWS BETWEEN unbounded preceding and current row),2) as cum_sal_expense from employees e
join departments d on e.department_id=d.department_id
where d.department_name in ('Finance','IT')
order by d.department_name,e.salary

 * postgresql://itv001477_hr_user:***@m01.itversity.com:5433/itv001477_hr_db
11 rows affected.


[(113, 'Finance', Decimal('6900.00'), Decimal('6900.00')),
 (111, 'Finance', Decimal('7700.00'), Decimal('14600.00')),
 (112, 'Finance', Decimal('7800.00'), Decimal('22400.00')),
 (110, 'Finance', Decimal('8200.00'), Decimal('30600.00')),
 (109, 'Finance', Decimal('9000.00'), Decimal('39600.00')),
 (108, 'Finance', Decimal('12000.00'), Decimal('51600.00')),
 (107, 'IT', Decimal('4200.00'), Decimal('4200.00')),
 (106, 'IT', Decimal('4800.00'), Decimal('9000.00')),
 (105, 'IT', Decimal('4800.00'), Decimal('13800.00')),
 (104, 'IT', Decimal('6000.00'), Decimal('19800.00')),
 (103, 'IT', Decimal('9000.00'), Decimal('28800.00'))]

### Exercise 3

Get top 3 paid employees with in each department by salary (use dense_rank)

* Use HR database employees and department tables for this problem.
* Highest paid employee should be ranked first.
* Output should contain employee_id, department_id, department_name, salary and employee_rank (derived field).
* Data should be sorted in ascending order by department_id in ascending order and then salary in descending order.

|employee_id|department_id|department_name|salary|employee_rank|
|---|---|---|---|---|
|200|10|Administration|4400.00|1|
|201|20|Marketing|13000.00|1|
|202|20|Marketing|6000.00|2|
|114|30|Purchasing|11000.00|1|
|115|30|Purchasing|3100.00|2|
|116|30|Purchasing|2900.00|3|
|203|40|Human Resources|6500.00|1|
|121|50|Shipping|8200.00|1|
|120|50|Shipping|8000.00|2|
|122|50|Shipping|7900.00|3|
|103|60|IT|9000.00|1|
|104|60|IT|6000.00|2|
|105|60|IT|4800.00|3|
|106|60|IT|4800.00|3|
|204|70|Public Relations|10000.00|1|
|145|80|Sales|14000.00|1|
|146|80|Sales|13500.00|2|
|147|80|Sales|12000.00|3|
|100|90|Executive|24000.00|1|
|101|90|Executive|17000.00|2|
|102|90|Executive|17000.00|2|
|108|100|Finance|12000.00|1|
|109|100|Finance|9000.00|2|
|110|100|Finance|8200.00|3|
|205|110|Accounting|12000.00|1|
|206|110|Accounting|8300.00|2|

In [5]:
%%sql
select * from(
    SELECT e.employee_id,e.department_id,d.department_name,e.salary,DENSE_RANK() OVER 
    (partition by d.department_name
     ORDER BY e.salary desc
    ) as employee_rank FROM employees e
    JOIN departments d on e.department_id=d.department_id
) q
where q.employee_rank <=3
order by q.department_id,q.salary desc

 * postgresql://itv001477_hr_user:***@m01.itversity.com:5433/itv001477_hr_db
26 rows affected.


[(200, 10, 'Administration', Decimal('4400.00'), 1),
 (201, 20, 'Marketing', Decimal('13000.00'), 1),
 (202, 20, 'Marketing', Decimal('6000.00'), 2),
 (114, 30, 'Purchasing', Decimal('11000.00'), 1),
 (115, 30, 'Purchasing', Decimal('3100.00'), 2),
 (116, 30, 'Purchasing', Decimal('2900.00'), 3),
 (203, 40, 'Human Resources', Decimal('6500.00'), 1),
 (121, 50, 'Shipping', Decimal('8200.00'), 1),
 (120, 50, 'Shipping', Decimal('8000.00'), 2),
 (122, 50, 'Shipping', Decimal('7900.00'), 3),
 (103, 60, 'IT', Decimal('9000.00'), 1),
 (104, 60, 'IT', Decimal('6000.00'), 2),
 (106, 60, 'IT', Decimal('4800.00'), 3),
 (105, 60, 'IT', Decimal('4800.00'), 3),
 (204, 70, 'Public Relations', Decimal('10000.00'), 1),
 (145, 80, 'Sales', Decimal('14000.00'), 1),
 (146, 80, 'Sales', Decimal('13500.00'), 2),
 (147, 80, 'Sales', Decimal('12000.00'), 3),
 (100, 90, 'Executive', Decimal('24000.00'), 1),
 (101, 90, 'Executive', Decimal('17000.00'), 2),
 (102, 90, 'Executive', Decimal('17000.00'), 2),
 (108, 100, 'Finance', Decimal('12000.00'), 1),
 (109, 100, 'Finance', Decimal('9000.00'), 2),
 (110, 100, 'Finance', Decimal('8200.00'), 3),
 (205, 110, 'Accounting', Decimal('12000.00'), 1),
 (206, 110, 'Accounting', Decimal('8300.00'), 2)]

### Exercise 4

Get top 3 products sold in the month of 2014 January by revenue.

* Use retail database tables such as orders, order_items and products.
* Consider only those orders which are either in **COMPLETE** or **CLOSED** status.
* Highest revenue generating product should come at top.
* Output should contain product_id, product_name, revenue, product_rank. **revenue** and **product_rank** are derived fields.
* Data should be sorted in descending order by revenue.

|product_id|product_name|revenue|product_rank|
|---|---|---|---|
|1004|Field & Stream Sportsman 16 Gun Fire Safe|250787.46|1|
|365|Perfect Fitness Perfect Rip Deck|151474.75|2|
|957|Diamondback Women's Serene Classic Comfort Bi|148190.12|3|


In [1]:
%load_ext sql

In [2]:
%env DATABASE_URL=postgresql://itv001477_retail_user:xndxklzpkk2h37en8z32vzeekbuvwkca@m01.itversity.com:5433/itv001477_retail_db

env: DATABASE_URL=postgresql://itv001477_retail_user:xndxklzpkk2h37en8z32vzeekbuvwkca@m01.itversity.com:5433/itv001477_retail_db


In [38]:
%%sql
select * from(
    select p.product_id,p.product_name,round(sum(oi.order_item_subtotal)::numeric,2) as revenue,
    ROW_NUMBER() over (order by round(sum(oi.order_item_subtotal)::numeric,2) desc) as product_rank
    from order_items oi
    join orders o on oi.order_item_order_id=o.order_id
    join products p on p.product_id=oi.order_item_product_id
    where to_char(o.order_date,'yyyy-MM') ~ '2014-01' and o.order_status in ('COMPLETE','CLOSED')
    group by p.product_id,p.product_name
) q
where q.product_rank <=3

 * postgresql://itv001477_retail_user:***@m01.itversity.com:5433/itv001477_retail_db
3 rows affected.


[(1004, 'Field & Stream Sportsman 16 Gun Fire Safe', Decimal('250787.46'), 1),
 (365, 'Perfect Fitness Perfect Rip Deck', Decimal('151474.75'), 2),
 (957, "Diamondback Women's Serene Classic Comfort Bi", Decimal('148190.12'), 3)]

### Exercise 5

Get top 3 products sold in the month of 2014 January under selected categories by revenue. The categories are **Cardio Equipment** and **Strength Training**.

* Use retail database tables such as orders, order_items, products as well as categories.
* Consider only those orders which are either in **COMPLETE** or **CLOSED** status.
* Highest revenue generating product should come at top.
* Output should contain category_id, category_name, product_id, product_name, revenue, product_rank. revenue and product_rank are derived fields.
* Data should be sorted in ascending order by category_id and descending order by revenue.

|category_id|category_name|product_id|product_name|revenue|product_rank|
|---|---|---|---|---|---|
|9|Cardio Equipment|191|Nike Men's Free 5.0+ Running Shoe|132286.77|1|
|9|Cardio Equipment|172|Nike Women's Tempo Shorts|870.00|2|
|10|Strength Training|208|SOLE E35 Elliptical|1999.99|1|
|10|Strength Training|203|GoPro HERO3+ Black Edition Camera|1199.97|2|
|10|Strength Training|216|Yakima DoubleDown Ace Hitch Mount 4-Bike Rack|189.00|3|

In [42]:
%%sql
select * from (
    select c.category_id,c.category_name,p.product_id,p.product_name,round(sum(oi.order_item_subtotal)::numeric,2) as revenue,
    DENSE_RANK() over (partition by c.category_name
                      order by round(sum(oi.order_item_subtotal)::numeric,2) desc) as product_rank
    from products p
    join order_items oi on oi.order_item_product_id=p.product_id
    join categories c on c.category_id=p.product_category_id
    join orders o on o.order_id=oi.order_item_order_id
    where o.order_status in ('COMPLETE','CLOSED') and c.category_name in ('Cardio Equipment','Strength Training') and
    to_char(o.order_date,'yyyy-MM') ~ '2014-01'
    group by c.category_id,c.category_name,p.product_id,p.product_name
) q
where q.product_rank <=3

 * postgresql://itv001477_retail_user:***@m01.itversity.com:5433/itv001477_retail_db
5 rows affected.


[(9, 'Cardio Equipment', 191, "Nike Men's Free 5.0+ Running Shoe", Decimal('132286.77'), 1),
 (9, 'Cardio Equipment', 172, "Nike Women's Tempo Shorts", Decimal('870.00'), 2),
 (10, 'Strength Training', 208, 'SOLE E35 Elliptical', Decimal('1999.99'), 1),
 (10, 'Strength Training', 203, 'GoPro HERO3+ Black Edition Camera', Decimal('1199.97'), 2),
 (10, 'Strength Training', 216, 'Yakima DoubleDown Ace Hitch Mount 4-Bike Rack', Decimal('189.00'), 3)]